converter/interceptor.go

```diff
@@ -0,0 +1,215 @@
+			err = decodePayloads(e.GetStartChildWorkflowExecutionInitiatedEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_SIGNAL_EXTERNAL_WORKFLOW_EXECUTION_INITIATED:
+			err = decodePayloads(e.GetSignalExternalWorkflowExecutionInitiatedEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_CONTINUED_AS_NEW:
+			err = decodePayloads(e.GetWorkflowExecutionContinuedAsNewEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_ACTIVITY_TASK_SCHEDULED:
+			err = decodePayloads(e.GetActivityTaskScheduledEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_ACTIVITY_TASK_COMPLETED:
+			err = decodePayloads(e.GetActivityTaskCompletedEventAttributes().Result, s.encoders...)
+		case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_SIGNALED:
+			err = decodePayloads(e.GetWorkflowExecutionSignaledEventAttributes().Input, s.encoders...)
+		case enumspb.EVENT_TYPE_CHILD_WORKFLOW_EXECUTION_COMPLETED:
+			err = decodePayloads(e.GetChildWorkflowExecutionCompletedEventAttributes().Result, s.encoders...)
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// NewPayloadEncoderGRPCServerInterceptor returns a GRPC Server Interceptor that will mimic the encoding
+// that the SDK system would perform when configured with a matching EncodingDataConverter.
+// Note: This approach does not support use cases that rely on the ContextAware DataConverter interface as
+// workflow context is not available at the GRPC level.
+func NewPayloadEncoderGRPCServerInterceptor(encoders ...PayloadEncoder) grpc.UnaryServerInterceptor {
+	s := serviceInterceptor{encoders: encoders}
+
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		err := s.processRequest(req)
+		if err != nil {
+			return nil, err
+		}
+
+		resp, err := handler(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+
+		return s.processResponse(resp), err
+	}
+}
+
+// NewPayloadEncoderGRPCClientInterceptor returns a GRPC Client Interceptor that will mimic the encoding
+// that the SDK system would perform when configured with a matching EncodingDataConverter.
+// Note: This approach does not support use cases that rely on the ContextAware DataConverter interface as
+// workflow context is not available at the GRPC level.
+func NewPayloadEncoderGRPCClientInterceptor(encoders ...PayloadEncoder) grpc.UnaryClientInterceptor {
+	s := serviceInterceptor{encoders: encoders}
+
+	return func(ctx context.Context, method string, req, response interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		err := s.processRequest(req)
+		if err != nil {
+			return err
+		}
+
+		err = invoker(ctx, method, req, response, cc, opts...)
+		if err != nil {
+			return err
+		}
+
+		return s.processResponse(response)
+	}
+}
+
+func encodePayloads(payloads *commonpb.Payloads, encoders ...PayloadEncoder) error {
+	for _, payload := range payloads.Payloads {
```

Comment:
You need to check whether `payloads` is nil here or call `payloads.GetPayloads()` instead